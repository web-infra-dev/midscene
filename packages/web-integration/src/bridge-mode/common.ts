export const DefaultBridgeServerPort = 3766;
export const DefaultLocalEndpoint = `http://127.0.0.1:${DefaultBridgeServerPort}`;
export const BridgeCallTimeout = 30000;

export enum BridgeEvent {
  Call = 'bridge-call',
  CallResponse = 'bridge-call-response',
  UpdateAgentStatus = 'bridge-update-agent-status',
  Message = 'bridge-message',
  Connected = 'bridge-connected',
  Refused = 'bridge-refused',
  ConnectNewTabWithUrl = 'connectNewTabWithUrl',
  ConnectCurrentTab = 'connectCurrentTab',
}

export interface BridgeConnectTabOptions {
  /**
   * If true, the page will always track the active tab.
   * @default true
   */
  forceSameTabNavigation?: boolean;
}

export enum MouseEvent {
  PREFIX = 'mouse.',
  Click = 'mouse.click',
  Wheel = 'mouse.wheel',
  Move = 'mouse.move',
  Drag = 'mouse.drag',
}

export enum KeyboardEvent {
  PREFIX = 'keyboard.',
  Type = 'keyboard.type',
  Press = 'keyboard.press',
}

export const BridgePageType = 'page-over-chrome-extension-bridge';

export const BridgeErrorCodeNoClientConnected = 'no-client-connected';

export interface BridgeCall {
  method: string;
  args: any[];
  response: any;
  callTime: number;
  responseTime: number;
  callback: (error: Error | undefined, response: any) => void;
  error?: Error;
}

export interface BridgeCallRequest {
  id: string;
  method: string;
  args: any[];
}

export interface BridgeCallResponse {
  id: string;
  response: any;
  error?: any;
}

export interface BridgeConnectedEventPayload {
  version: string;
}
