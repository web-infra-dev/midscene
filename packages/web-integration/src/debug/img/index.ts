import assert from 'node:assert';
import { Buffer } from 'node:buffer';
import type { NodeType } from '@/extractor/constants';
import sharp from 'sharp';

// Define picture path
type ElementType = {
  x: number;
  y: number;
  width: number;
  height: number;
  label: string;
  attributes: {
    [key: string]: string;
    nodeType: NodeType;
  };
};

const createSvgOverlay = (
  elements: Array<ElementType>,
  imageWidth: number,
  imageHeight: number,
) => {
  let svgContent = `<svg width="${imageWidth}" height="${imageHeight}" xmlns="http://www.w3.org/2000/svg">`;

  // Define color array
  const colors = [
    { rect: 'blue', text: 'white' },
    { rect: 'green', text: 'white' },
  ];

  // Define clipping path
  svgContent += '<defs>';
  elements.forEach((element, index) => {
    svgContent += `
      <clipPath id="clip${index}">
        <rect x="${element.x}" y="${element.y}" width="${element.width}" height="${element.height}" />
      </clipPath>
    `;
  });
  svgContent += '</defs>';

  elements.forEach((element, index) => {
    // Calculate the width and height of the text
    const textWidth = element.label.length * 8; // Assume that each character is 8px wide
    const textHeight = 12; // Assume that the text height is 20px

    // Calculates the position of the initial color block so that it wraps and centers the text
    const rectWidth = textWidth + 5;
    const rectHeight = textHeight + 4;
    let rectX = element.x - rectWidth;
    let rectY = element.y + element.height / 2 - textHeight / 2 - 2;

    // Initial text position
    let textX = rectX + rectWidth / 2;
    let textY = rectY + rectHeight / 2 + 6;

    // Check to see if it's obscured by the left
    if (rectX < 0) {
      rectX = element.x;
      rectY = element.y - rectHeight;
      textX = rectX + rectWidth / 2;
      textY = rectY + rectHeight / 2 + 6;
    }

    // Choose color
    const color = colors[index % colors.length];

    // Draw boxes and text
    svgContent += `
      <rect x="${element.x}" y="${element.y}" width="${element.width}" height="${element.height}" 
            style="fill:none;stroke:${color.rect};stroke-width:4" clip-path="url(#clip${index})" />
      <rect x="${rectX}" y="${rectY}" width="${rectWidth}" height="${rectHeight}" style="fill:${color.rect};" />
      <text x="${textX}" y="${textY}" 
            text-anchor="middle" dominant-baseline="middle" style="fill:${color.text};font-size:12px;font-weight:bold;">
        ${element.label}
      </text>
    `;
  });

  svgContent += '</svg>';
  return Buffer.from(svgContent);
};

export const processImageElementInfo = async (options: {
  inputImgBase64: string;
  elementsPositionInfo: Array<ElementType>;
  elementsPositionInfoWithoutText: Array<ElementType>;
}) => {
  // Get the size of the original image
  const base64Image = options.inputImgBase64.split(';base64,').pop();
  assert(base64Image, 'base64Image is undefined');

  const imageBuffer = Buffer.from(base64Image, 'base64');
  const metadata = await sharp(imageBuffer).metadata();
  const { width, height } = metadata;

  if (width && height) {
    // Create svg overlay
    const svgOverlay = createSvgOverlay(
      options.elementsPositionInfo,
      width,
      height,
    );
    const svgOverlayWithoutText = createSvgOverlay(
      options.elementsPositionInfoWithoutText,
      width,
      height,
    );

    // Composite picture
    const compositeElementInfoImgBase64 = await sharp(imageBuffer)
      // .resize(newDimensions.width, newDimensions.height)
      .composite([{ input: svgOverlay, blend: 'over' }])
      .toBuffer()
      .then((data) => {
        // Convert image data to base64 encoding
        return data.toString('base64');
      })
      .catch((err) => {
        throw err;
      });

    // Composite picture withoutText
    const compositeElementInfoImgWithoutTextBase64 = await sharp(imageBuffer)
      // .resize(newDimensions.width, newDimensions.height)
      .composite([{ input: svgOverlayWithoutText, blend: 'over' }])
      .toBuffer()
      .then((data) => {
        // Convert image data to base64 encoding
        return data.toString('base64');
      })
      .catch((err) => {
        throw err;
      });

    return {
      compositeElementInfoImgBase64,
      compositeElementInfoImgWithoutTextBase64,
    };
  }
  throw Error('Image processing failed because width or height is undefined');
};

export const compressImageSize = async (buffer: Buffer) => {
  const resizedBuffer = await sharp(buffer).resize({ width: 640 }).toBuffer();

  return resizedBuffer;
};
