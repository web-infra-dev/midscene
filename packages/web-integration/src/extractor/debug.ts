import { webExtractTextWithPosition } from '.';
import { setExtractTextWithPositionOnWindow } from './util';

console.log(webExtractTextWithPosition(document.body, true));
console.log(JSON.stringify(webExtractTextWithPosition(document.body, true)));
setExtractTextWithPositionOnWindow();
