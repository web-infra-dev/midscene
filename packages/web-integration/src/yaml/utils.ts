import assert from 'node:assert';
import yaml from 'js-yaml';

import type { MidsceneYamlScript } from '@midscene/core';
import type {
  MidsceneYamlFlowItem,
  MidsceneYamlFlowItemAIAction,
  MidsceneYamlFlowItemAIAssert,
  MidsceneYamlFlowItemAIQuery,
  MidsceneYamlFlowItemAIWaitFor,
  MidsceneYamlFlowItemSleep,
} from '@midscene/core';

function interpolateEnvVars(content: string): string {
  return content.replace(/\$\{([^}]+)\}/g, (_, envVar) => {
    const value = process.env[envVar.trim()];
    if (value === undefined) {
      throw new Error(`Environment variable "${envVar.trim()}" is not defined`);
    }
    return value;
  });
}

export function parseYamlScript(
  content: string,
  filePath?: string,
): MidsceneYamlScript {
  const interpolatedContent = interpolateEnvVars(content);
  const obj = yaml.load(interpolatedContent) as MidsceneYamlScript;
  const pathTip = filePath ? `, failed to load ${filePath}` : '';
  assert(obj.target, `property "target" is required in yaml script${pathTip}`);
  assert(
    typeof obj.target === 'object',
    `property "target" must be an object${pathTip}`,
  );
  assert(obj.tasks, `property "tasks" is required in yaml script${pathTip}`);
  assert(
    Array.isArray(obj.tasks),
    `property "tasks" must be an array${pathTip}`,
  );
  return obj;
}

export const flowItemBrief = (flowItem?: MidsceneYamlFlowItem) => {
  if (!flowItem) {
    return '';
  }

  const sliceText = (text?: string) => {
    const lengthLimit = 60;
    if (text && text.length > lengthLimit) {
      return `${text.slice(0, lengthLimit)}...`;
    }

    return text || '';
  };

  if (
    (flowItem as MidsceneYamlFlowItemAIAction).aiAction ||
    (flowItem as MidsceneYamlFlowItemAIAction).ai
  ) {
    const lastTip = (
      (flowItem as MidsceneYamlFlowItemAIAction).aiActionProgressTips || []
    ).at(-1);
    return `aiAction: ${sliceText(
      lastTip ||
        (flowItem as MidsceneYamlFlowItemAIAction).aiAction ||
        (flowItem as MidsceneYamlFlowItemAIAction).ai,
    )}`;
  }
  if ((flowItem as MidsceneYamlFlowItemAIAssert).aiAssert) {
    return `aiAssert: ${sliceText(
      (flowItem as MidsceneYamlFlowItemAIAssert).aiAssert,
    )}`;
  }
  if ((flowItem as MidsceneYamlFlowItemAIQuery).aiQuery) {
    return `aiQuery: ${sliceText((flowItem as MidsceneYamlFlowItemAIQuery).aiQuery)}`;
  }
  if ((flowItem as MidsceneYamlFlowItemAIWaitFor).aiWaitFor) {
    return `aiWaitFor: ${sliceText(
      (flowItem as MidsceneYamlFlowItemAIWaitFor).aiWaitFor,
    )}`;
  }
  if ((flowItem as MidsceneYamlFlowItemSleep).sleep) {
    return `sleep: ${(flowItem as MidsceneYamlFlowItemSleep).sleep}`;
  }
  return '';
};
