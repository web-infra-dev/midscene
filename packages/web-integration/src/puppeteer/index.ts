import { PageAgent, type PageAgentOpt } from '@/common/agent';
import type { Page as PuppeteerPage } from 'puppeteer';
import { WebPage as PuppeteerWebPage } from './page';

export { WebPage as PuppeteerWebPage } from './page';
export class PuppeteerAgent extends PageAgent {
  constructor(page: PuppeteerPage, opts?: PageAgentOpt) {
    const webPage = new PuppeteerWebPage(page);
    super(webPage, opts);
  }
}
