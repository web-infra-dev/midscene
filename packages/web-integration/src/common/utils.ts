import assert from 'node:assert';
import type { Buffer } from 'node:buffer';
import { randomUUID } from 'node:crypto';
import { readFileSync } from 'node:fs';
import path from 'node:path';
import type { ElementInfo } from '@/extractor';
import type { PlaywrightParserOpt, UIContext } from '@midscene/core';
import { getTmpFile } from '@midscene/core/utils';
import { findNearestPackageJson } from '@midscene/shared/fs';
import { base64Encoded, imageInfoOfBase64 } from '@midscene/shared/img';
import dayjs from 'dayjs';
import { WebElementInfo } from '../web-element';
import type { WebPage } from './page';

export type WebUIContext = UIContext<WebElementInfo> & {
  url: string;
};

export async function parseContextFromWebPage(
  page: WebPage,
  _opt?: PlaywrightParserOpt,
): Promise<WebUIContext> {
  assert(page, 'page is required');

  const url = page.url();
  const file = await page.screenshot();
  const screenshotBuffer = readFileSync(file);
  const screenshotBase64 = base64Encoded(file);
  const captureElementSnapshot = await page.getElementInfos();

  // align element
  const elementsInfo = await alignElements(
    screenshotBuffer,
    captureElementSnapshot,
    page,
  );

  const size = await imageInfoOfBase64(screenshotBase64);

  return {
    content: elementsInfo,
    size,
    screenshotBase64,
    url,
  };
}

export async function getExtraReturnLogic() {
  const pathDir = findNearestPackageJson(__dirname);
  assert(pathDir, `can't find pathDir, with ${__dirname}`);
  const scriptPath = path.join(pathDir, './dist/script/htmlElement.js');
  const elementInfosScriptContent = readFileSync(scriptPath, 'utf-8');
  return `${elementInfosScriptContent}midscene_element_inspector.webExtractTextWithPosition()`;
}

const sizeThreshold = 3;
async function alignElements(
  screenshotBuffer: Buffer,
  elements: ElementInfo[],
  page: WebPage,
): Promise<WebElementInfo[]> {
  const validElements = elements.filter((item) => {
    return (
      item.rect.height >= sizeThreshold && item.rect.width >= sizeThreshold
    );
  });
  const textsAligned: WebElementInfo[] = [];
  for (const item of validElements) {
    const { rect, id, content, attributes, locator } = item;
    textsAligned.push(
      new WebElementInfo({
        rect,
        locator,
        id,
        content,
        attributes,
        page,
      }),
    );
  }

  return textsAligned;
}

export function reportFileName(tag = 'web') {
  const dateTimeInFileName = dayjs().format('YYYY-MM-DD_HH-mm-ss-SSS');
  return `${tag}-${dateTimeInFileName}`;
}

export function printReportMsg(filepath: string) {
  console.log('Midscene - report file updated:', filepath);
}

/**
 * Get the current execution file name
 * @returns The name of the current execution file
 */
export function getCurrentExecutionFile(trace?: string): string | false {
  const error = new Error();
  const stackTrace = trace || error.stack;
  const pkgDir = process.cwd() || '';
  if (stackTrace) {
    const stackLines = stackTrace.split('\n');
    for (const line of stackLines) {
      if (
        line.includes('.spec.') ||
        line.includes('.test.') ||
        line.includes('.ts') ||
        line.includes('.js')
      ) {
        const match = line.match(/(?:at\s+)?(.*?\.(?:spec|test)\.[jt]s)/);
        if (match?.[1]) {
          const targetFileName = match[1]
            .replace(pkgDir, '')
            .trim()
            .replace('at ', '');
          return targetFileName;
        }
      }
    }
  }
  return false;
}

/**
 * Generates a unique cache ID based on the current execution file and a counter.
 *
 * This function creates a cache ID by combining the name of the current execution file
 * (typically a test or spec file) with an incrementing index. This ensures that each
 * cache ID is unique within the context of a specific test file across multiple executions.
 *
 * The function uses a Map to keep track of the index for each unique file, incrementing
 * it with each call for the same file.
 *
 * @returns {string} A unique cache ID in the format "filename-index"
 *
 * @example
 * // First call for "example.spec.ts"
 * generateCacheId(); // Returns "example.spec.ts-1"
 *
 * // Second call for "example.spec.ts"
 * generateCacheId(); // Returns "example.spec.ts-2"
 *
 * // First call for "another.test.ts"
 * generateCacheId(); // Returns "another.test.ts-1"
 */

const testFileIndex = new Map<string, number>();
export function generateCacheId(fileName?: string): string {
  let taskFile = fileName || getCurrentExecutionFile();
  if (!taskFile) {
    taskFile = randomUUID();
    console.warn(
      'Midscene - using random UUID for cache id. Cache may be invalid.',
    );
  }

  if (testFileIndex.has(taskFile)) {
    const currentIndex = testFileIndex.get(taskFile);
    if (currentIndex !== undefined) {
      testFileIndex.set(taskFile, currentIndex + 1);
    }
  } else {
    testFileIndex.set(taskFile, 1);
  }
  return `${taskFile}-${testFileIndex.get(taskFile)}`;
}
