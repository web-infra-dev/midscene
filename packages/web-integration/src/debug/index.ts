import { existsSync, mkdirSync, readFileSync, writeFileSync } from 'node:fs';
import path from 'node:path';
import type { WebPage } from '@/common/page';
import type { ElementInfo } from '@/extractor';
import { NodeType } from '@/extractor/constants';
import {
  processImageElementInfo,
  resizeImg,
  saveBase64Image,
} from '@midscene/shared/img';

export async function generateExtractData(
  page: WebPage,
  targetDir: string,
  saveImgType?: {
    disableInputImage: boolean;
    disableOutputImage: boolean;
    disableOutputWithoutTextImg: boolean;
    disableResizeOutputImg: boolean;
    disableSnapshot: boolean;
  },
) {
  const file = await page.screenshot();
  const screenshotBuffer = readFileSync(file);

  const inputImgBase64 = screenshotBuffer.toString('base64');

  const {
    elementsPositionInfo,
    captureElementSnapshot,
    elementsPositionInfoWithoutText,
  } = await getElementInfos(page);

  const inputImagePath = path.join(targetDir, 'input.png');
  const outputImagePath = path.join(targetDir, 'output.png');
  const outputWithoutTextImgPath = path.join(
    targetDir,
    'output_without_text.png',
  );
  const resizeOutputImgPath = path.join(targetDir, 'resize-output.png');
  const snapshotJsonPath = path.join(targetDir, 'element-snapshot.json');

  const {
    compositeElementInfoImgBase64,
    compositeElementInfoImgWithoutTextBase64,
  } = await processImageElementInfo({
    elementsPositionInfo,
    elementsPositionInfoWithoutText,
    inputImgBase64,
  });

  const resizeImgBase64 = (await resizeImg(inputImgBase64)) as string;

  const existingSnapshot = existsSync(snapshotJsonPath)
    ? JSON.parse(readFileSync(snapshotJsonPath, 'utf-8'))
    : null;

  if (
    existingSnapshot &&
    JSON.stringify(existingSnapshot) === JSON.stringify(captureElementSnapshot)
  ) {
    console.log('skip save snapshot for ', targetDir);
    return;
  }

  if (!saveImgType?.disableSnapshot) {
    writeFileSyncWithDir(
      snapshotJsonPath,
      JSON.stringify(captureElementSnapshot, null, 2),
    );
  }
  if (!saveImgType?.disableInputImage) {
    await saveBase64Image({
      base64Data: inputImgBase64,
      outputPath: inputImagePath,
    });
  }
  if (!saveImgType?.disableOutputImage) {
    await saveBase64Image({
      base64Data: compositeElementInfoImgBase64,
      outputPath: outputImagePath,
    });
  }
  if (!saveImgType?.disableOutputWithoutTextImg) {
    await saveBase64Image({
      base64Data: compositeElementInfoImgWithoutTextBase64,
      outputPath: outputWithoutTextImgPath,
    });
  }
  if (!saveImgType?.disableResizeOutputImg) {
    await saveBase64Image({
      base64Data: resizeImgBase64,
      outputPath: resizeOutputImgPath,
    });
  }
}

export function generateTestDataPath(testDataName: string) {
  // `dist/lib/index.js` Is the default export path
  const modulePath = require
    .resolve('@midscene/core')
    .replace('dist/lib/index.js', '');
  const midsceneTestDataPath = path.join(
    modulePath,
    `tests/ai/evaluate/test-data/${testDataName}`,
  );

  return midsceneTestDataPath;
}

function ensureDirectoryExistence(filePath: string) {
  const dirname = path.dirname(filePath);
  if (existsSync(dirname)) {
    return;
  }
  ensureDirectoryExistence(dirname);
  mkdirSync(dirname);
}

type WriteFileSyncParams = Parameters<typeof writeFileSync>;

export function writeFileSyncWithDir(
  filePath: string,
  content: WriteFileSyncParams[1],
  options: WriteFileSyncParams[2] = {},
) {
  ensureDirectoryExistence(filePath);
  writeFileSync(filePath, content, options);
}

export async function getElementInfos(page: WebPage) {
  const captureElementSnapshot: Array<ElementInfo> =
    await page.getElementInfos();

  const elementsPositionInfo = captureElementSnapshot.map(
    (elementInfo, index) => {
      return {
        label: elementInfo.indexId?.toString() || index.toString(),
        x: elementInfo.rect.left,
        y: elementInfo.rect.top,
        width: elementInfo.rect.width,
        height: elementInfo.rect.height,
        attributes: elementInfo.attributes,
      };
    },
  );
  const elementsPositionInfoWithoutText = elementsPositionInfo.filter(
    (elementInfo) => {
      if (elementInfo.attributes.nodeType === NodeType.TEXT) {
        return false;
      }
      return true;
    },
  );
  return {
    elementsPositionInfo,
    captureElementSnapshot,
    elementsPositionInfoWithoutText,
  };
}
