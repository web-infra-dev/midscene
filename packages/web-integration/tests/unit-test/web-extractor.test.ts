import path, { join } from 'node:path';
import { parseContextFromWebPage } from '@/common/utils';
import { generateExtractData } from '@/debug';
import { imageInfo } from '@midscene/shared/img';
import { describe, expect, it } from 'vitest';
import { launchPage } from '../ai/web/puppeteer/utils';

const pagePath = join(__dirname, './fixtures/web-extractor/index.html');
describe(
  'extractor',
  () => {
    it('basic', async () => {
      const { page, reset } = await launchPage(`file://${pagePath}`, {
        viewport: {
          width: 1080,
          height: 2000,
        },
      });

      const { content } = await parseContextFromWebPage(page);
      await generateExtractData(
        page,
        path.join(__dirname, 'fixtures/web-extractor'),
        {
          disableInputImage: false,
          disableOutputImage: false,
          disableOutputWithoutTextImg: true,
          disableResizeOutputImg: true,
          disableSnapshot: true,
        },
      );

      const list = content.map((item) => {
        return {
          content: item.content,
          attributes: item.attributes,
        };
      });

      expect(list).toMatchSnapshot();
      await reset();
    });

    it('check screenshot size - 1x', async () => {
      const { page, reset } = await launchPage(`file://${pagePath}`, {
        viewport: {
          width: 1080,
          height: 2000,
          deviceScaleFactor: 1,
        },
      });

      const shotpath = await page.screenshot();

      const info = await imageInfo(shotpath);
      expect(info.width).toBe(1080);
      expect(info.height).toBe(2000);
      await reset();
    });

    it('check screenshot size - 2x', async () => {
      const { page, reset } = await launchPage(`file://${pagePath}`, {
        viewport: {
          width: 1080,
          height: 2000,
          deviceScaleFactor: 2,
        },
      });

      const shotpath = await page.screenshot();

      const info = await imageInfo(shotpath);
      expect(info.width).toBe(1080); // always 1x for screenshot
      expect(info.height).toBe(2000);
      await reset();
    });

    it('scroll', async () => {
      const { page, reset } = await launchPage(`file://${pagePath}`, {
        viewport: {
          width: 1080,
          height: 200,
        },
      });
      await page.scrollDownOneScreen();
      await new Promise((resolve) => setTimeout(resolve, 1000));
      await generateExtractData(
        page,
        path.join(__dirname, 'fixtures/web-extractor/scroll'),
        {
          disableInputImage: false,
          disableOutputImage: false,
          disableOutputWithoutTextImg: true,
          disableResizeOutputImg: true,
          disableSnapshot: true,
        },
      );
    });

    it('profile ', async () => {
      const { page, reset } = await launchPage('https://webinfra.org/about');
      await new Promise((resolve) => setTimeout(resolve, 1000));
      console.time('total - parseContextFromWebPage');
      const { content } = await parseContextFromWebPage(page);
      console.timeEnd('total - parseContextFromWebPage');
      await reset();
    });
  },
  {
    timeout: 90 * 1000,
  },
);
