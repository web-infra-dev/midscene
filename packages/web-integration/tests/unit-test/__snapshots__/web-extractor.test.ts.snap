// Vitest Snapshot v1, https://vitest.dev/guide/snapshot.html

exports[`extractor > basic 1`] = `
[
  {
    "attributes": {
      "nodeType": "TEXT Node",
    },
    "content": "Data Record",
  },
  {
    "attributes": {
      "nodeType": "TEXT Node",
    },
    "content": "1970-01-01 19:25:01",
  },
  {
    "attributes": {
      "nodeType": "TEXT Node",
    },
    "content": "User Name: Stella",
  },
  {
    "attributes": {
      "nodeType": "TEXT Node",
    },
    "content": "ID",
  },
  {
    "attributes": {
      "nodeType": "TEXT Node",
    },
    "content": "Field 2",
  },
  {
    "attributes": {
      "nodeType": "TEXT Node",
    },
    "content": "Field 3",
  },
  {
    "attributes": {
      "nodeType": "TEXT Node",
    },
    "content": "Field 4",
  },
  {
    "attributes": {
      "nodeType": "TEXT Node",
    },
    "content": "Field 5",
  },
  {
    "attributes": {
      "nodeType": "TEXT Node",
    },
    "content": "30S",
  },
  {
    "attributes": {
      "nodeType": "TEXT Node",
    },
    "content": "Kace Cervantes",
  },
  {
    "attributes": {
      "nodeType": "TEXT Node",
    },
    "content": "Aylin Sawyer",
  },
  {
    "attributes": {
      "nodeType": "TEXT Node",
    },
    "content": "Jefferson Kirby",
  },
  {
    "attributes": {
      "nodeType": "TEXT Node",
    },
    "content": "Skyla Jefferson",
  },
  {
    "attributes": {
      "nodeType": "TEXT Node",
    },
    "content": "70U",
  },
  {
    "attributes": {
      "nodeType": "TEXT Node",
    },
    "content": "Florence Davenport",
  },
  {
    "attributes": {
      "nodeType": "TEXT Node",
    },
    "content": "Dariel Acevedo",
  },
  {
    "attributes": {
      "nodeType": "TEXT Node",
    },
    "content": "Ashlynn Delacruz",
  },
  {
    "attributes": {
      "nodeType": "TEXT Node",
    },
    "content": "Memphis Leal",
  },
  {
    "attributes": {
      "nodeType": "TEXT Node",
    },
    "content": "3AY",
  },
  {
    "attributes": {
      "nodeType": "TEXT Node",
    },
    "content": "Crystal Newman",
  },
  {
    "attributes": {
      "nodeType": "TEXT Node",
    },
    "content": "Anderson Brown",
  },
  {
    "attributes": {
      "nodeType": "TEXT Node",
    },
    "content": "Charlotte Griffith",
  },
  {
    "attributes": {
      "nodeType": "TEXT Node",
    },
    "content": "Franklin Everett",
  },
  {
    "attributes": {
      "nodeType": "TEXT Node",
    },
    "content": "YPG",
  },
  {
    "attributes": {
      "nodeType": "TEXT Node",
    },
    "content": "Kori Payne",
  },
  {
    "attributes": {
      "nodeType": "TEXT Node",
    },
    "content": "Edward Blevins",
  },
  {
    "attributes": {
      "nodeType": "TEXT Node",
    },
    "content": "Aila Gill",
  },
  {
    "attributes": {
      "nodeType": "TEXT Node",
    },
    "content": "Matthias Reed",
  },
  {
    "attributes": {
      "nodeType": "TEXT Node",
    },
    "content": "ZEN",
  },
  {
    "attributes": {
      "nodeType": "TEXT Node",
    },
    "content": "Magnolia Duke",
  },
  {
    "attributes": {
      "nodeType": "TEXT Node",
    },
    "content": "Kalel Glover",
  },
  {
    "attributes": {
      "nodeType": "TEXT Node",
    },
    "content": "Alessia Barton",
  },
  {
    "attributes": {
      "nodeType": "TEXT Node",
    },
    "content": "Cassius Peck",
  },
  {
    "attributes": {
      "nodeType": "TEXT Node",
    },
    "content": "Form",
  },
  {
    "attributes": {
      "nodeType": "TEXT Node",
    },
    "content": "Name:",
  },
  {
    "attributes": {
      "htmlTagName": "<input>",
      "id": "J_input",
      "nodeType": "FORM_ITEM Node",
      "placeholder": "Hello World This is Placeholder",
    },
    "content": "Hello World This is Placeholder",
  },
  {
    "attributes": {
      "htmlTagName": "<input>",
      "id": "J_input",
      "nodeType": "FORM_ITEM Node",
      "placeholder": "Hello World This is Placeholder",
      "value": "Now I am a value, instead of placeholder",
    },
    "content": "Now I am a value, instead of placeholder",
  },
  {
    "attributes": {
      "nodeType": "BUTTON Node",
    },
    "content": "Click Me",
  },
  {
    "attributes": {
      "nodeType": "TEXT Node",
    },
    "content": "Shape",
  },
  {
    "attributes": {
      "htmlTagName": "<input>",
      "nodeType": "FORM_ITEM Node",
      "placeholder": "You shouldn't see this placeholder.",
      "value": "Rectangle",
    },
    "content": "Rectangle",
  },
  {
    "attributes": {
      "htmlTagName": "<textarea>",
      "nodeType": "FORM_ITEM Node",
      "placeholder": "this_is_a_textarea",
    },
    "content": "this_is_a_textarea",
  },
  {
    "attributes": {
      "nodeType": "TEXT Node",
    },
    "content": "English",
  },
  {
    "attributes": {
      "nodeType": "TEXT Node",
    },
    "content": "中文",
  },
  {
    "attributes": {
      "nodeType": "TEXT Node",
    },
    "content": "Tiếng Việt",
  },
  {
    "attributes": {
      "nodeType": "TEXT Node",
    },
    "content": "Choose an option:",
  },
  {
    "attributes": {
      "htmlTagName": "<select>",
      "id": "options",
      "name": "options",
      "nodeType": "FORM_ITEM Node",
    },
    "content": "Option 1",
  },
  {
    "attributes": {
      "nodeType": "TEXT Node",
    },
    "content": "This is zoomed content",
  },
  {
    "attributes": {
      "nodeType": "TEXT Node",
    },
    "content": "Something Else",
  },
  {
    "attributes": {
      "nodeType": "TEXT Node",
    },
    "content": "输入搜索关键词",
  },
  {
    "attributes": {
      "aria-autocomplete": "list",
      "aria-expanded": "false",
      "aria-haspopup": "true",
      "aria-label": "搜索任何物品",
      "aria-owns": "ui-id-1",
      "autocapitalize": "off",
      "autocomplete": "off",
      "autocorrect": "off",
      "class": ".gh-tb.ui-autocomplete-input",
      "htmlTagName": "<input>",
      "id": "gh-ac",
      "maxlength": "300",
      "name": "_nkw",
      "nodeType": "FORM_ITEM Node",
      "placeholder": "搜索任何物品",
      "role": "combobox",
      "size": "50",
      "spellcheck": "false",
      "type": "text",
    },
    "content": "搜索任何物品",
  },
  {
    "attributes": {
      "class": ".life-core-input.life-core-input-size-md",
      "htmlTagName": "<input>",
      "nodeType": "FORM_ITEM Node",
      "placeholder": "验证码",
      "tabindex": "0",
      "type": "text",
    },
    "content": "验证码",
  },
  {
    "attributes": {
      "class": ".life-core-checkbox-icon",
      "nodeType": "CONTAINER Node",
    },
    "content": "",
  },
  {
    "attributes": {
      "aria-label": "Search",
      "class": ".btn",
      "nodeType": "BUTTON Node",
      "tabindex": "0",
      "type": "submit",
    },
    "content": " ",
  },
  {
    "attributes": {
      "nodeType": "TEXT Node",
    },
    "content": "content 000",
  },
  {
    "attributes": {
      "nodeType": "CONTAINER Node",
      "style": "height: 30px;width: 100px;",
    },
    "content": "",
  },
  {
    "attributes": {
      "nodeType": "TEXT Node",
    },
    "content": "content AAA",
  },
  {
    "attributes": {
      "nodeType": "IMG Node",
      "src": "data:image/svg+xml;...",
    },
    "content": "",
  },
  {
    "attributes": {
      "nodeType": "BUTTON Node",
      "very-long-attr": "width: 100px; height: 100px; background-color: #cc...",
    },
    "content": "long-style-content",
  },
  {
    "attributes": {
      "nodeType": "TEXT Node",
    },
    "content": "Click me",
  },
  {
    "attributes": {
      "aria-controls": "semi-select-5yxiyng",
      "class": ".widget",
      "nodeType": "CONTAINER Node",
    },
    "content": "",
  },
  {
    "attributes": {
      "aria-labelledby": "eval_object.object_type-label",
      "class": ".widget",
      "nodeType": "CONTAINER Node",
    },
    "content": "",
  },
  {
    "attributes": {
      "nodeType": "TEXT Node",
    },
    "content": "Content 1",
  },
  {
    "attributes": {
      "class": ".child-container-without-content-2-1",
      "nodeType": "CONTAINER Node",
      "style": "width: 100px; height: 100px; background-color: #cc...",
    },
    "content": "",
  },
  {
    "attributes": {
      "nodeType": "TEXT Node",
    },
    "content": "Content 2",
  },
  {
    "attributes": {
      "nodeType": "TEXT Node",
    },
    "content": "Nested Content 3",
  },
  {
    "attributes": {
      "class": ".child-container-4",
      "nodeType": "CONTAINER Node",
      "style": "width: 100px; height: 100px; background-color: #cc...",
    },
    "content": "",
  },
  {
    "attributes": {
      "nodeType": "IMG Node",
      "svgContent": "true",
    },
    "content": "",
  },
  {
    "attributes": {
      "frameborder": "0",
      "height": "400px",
      "nodeType": "CONTAINER Node",
      "src": "child.html",
      "width": "400px",
    },
    "content": "",
  },
]
`;
